### ISSUE TITLE
Producer create method blocked forever 

### ISSUE THREAD

**Issue by tcbuzor** [apache/pulsar#4988]
**Describe the bug**
When we call the create method from the producer and there are no partitions available in the broker, the create method never returns and the thread blocks trying to connect. The connectionTimeout does not seem to help. The only option is to restart the entire JVM

**To Reproduce**
Try to create a producer on a broker with insufficient partition.

**Expected behavior**
The producer should return after the timeout (connectionTimeout). Right now, the connectionTimeout does not have any effect. I do not have see anywhere in the code where the connectionTimeout is been used.

**Screenshots**

**Desktop (please complete the following information):**

**Additional context**

**Comment by wolfstudy** [comment]
> there are no partitions available in the broker,

By default, when the used topic is not created, the specified topic is automatically created when the producer is created. We can modify this behavior in `broker.conf`:

```
# Enable topic auto creation if new producer or consumer connected (disable auto creation with value false)
allowAutoTopicCreation=true
```

Can you provide a simple code to stabilize this problem?

**Comment by tcbuzor** [comment]
When a Producer.create is called, how long does it create request last ? does it wait forever for the producer to be created? Can I pass a timeout to the create method ?
I dont have to restart the JVM to remove blocking threads due to hanging Producer create call.

**Comment by vivekkrj** [comment]
Here is the sample code base for the client and pruducer block:

/**
	 * 
	 * @throws PulsarClientException
	 */
	private PulsarClient getProducerClient() throws PulsarClientException {
		PulsarClient client = null;
		if(producerOptionBean.isAuthenticationEnable()) {
			Authentication authentication = AuthenticationFactory
					.TLS(producerOptionBean.getClientCertificateFile(), producerOptionBean.getClientCertificateKey());
			client = PulsarClient
					.builder()
					.serviceUrl(producerOptionBean.getPulsarBrokerUrl())
					.tlsTrustCertsFilePath(producerOptionBean.getTrustCa())
					.allowTlsInsecureConnection(producerOptionBean.isAllowTlsInsecureConnection())
					.enableTlsHostnameVerification(false)
					.authentication(authentication)
					.build();
		}
		else {
			client = PulsarClient
					.builder()
					.serviceUrl(producerOptionBean.getPulsarBrokerUrl())
					.tlsTrustCertsFilePath(producerOptionBean.getTrustCa())
					.allowTlsInsecureConnection(producerOptionBean.isAllowTlsInsecureConnection())
					.build();
		}
		return client;
	}

/**
	 * 
	 * @throws PulsarClientException
	 */
	private Producer<byte[]> buildProducer(PulsarClient client, String producerName) throws PulsarClientException {
		Producer<byte[]> producer = client.newProducer()
				.topic(producerOptionBean.getPulsarTopic())
				.producerName(producerName)
				.enableBatching(producerOptionBean.isEnableBatching())
				.batchingMaxPublishDelay(producerOptionBean.getBatchMessageMaxDelay(), TimeUnit.MILLISECONDS)
				.sendTimeout(producerOptionBean.getMsgAckTimeout(), TimeUnit.SECONDS)
				.compressionType(getCompressionType())
				.maxPendingMessages(producerOptionBean.getMaxPendingMessages())
				.batchingMaxMessages(producerOptionBean.getBatchingMaxMessages())
				.blockIfQueueFull(producerOptionBean.isBlockIfQueueFull())
				.create();
		return producer;
	}

**Comment by vivekkrj** [comment]
Here is the JSTACK of the JVm where you can see that thread is stuck for forver in-order to create to be completed:

"Streaming.Test: netty server 5" #99780 prio=5 os_prio=0 tid=0x00002b41d431a800 nid=0x6929 waiting on condition [0x00002b40a1cda000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000071d8d0308> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
	at org.apache.pulsar.client.impl.ProducerBuilderImpl.create(ProducerBuilderImpl.java:86)

**Comment by vivekkrj** [comment]
@wolfstudy  : I have provided above both buliding producer and JSTACK where thread is stuck on create method. Let me know if you need anything else.

**Comment by whisper-bye** [comment]
+1
Failed to create Producer: ProducerBlockedQuotaExceededError

**Comment by cerebrotecnologico** [comment]
As a workaroud, I used ```createAsync()```

```
try {
        f =
            (CompletableFuture<Producer<?>>)
                pc.newProducer(...)
                     ...some other config of the producer...
                    .createAsync();
        return f.get(maxWaitCreateProducerMillis, TimeUnit.MILLISECONDS);
      } catch (TimeoutException e) {
        if (f != null) {
          f.cancel(true);
        }
        throw e;
      }
```

**Comment by vivekkrj** [comment]
I did solve this long time back using a different design model.

**Comment by cerebrotecnologico** [comment]
> I did solve this long time back using a different design model.

Would you mind sharing it?

**Comment by tisonkun** [comment]
Closing as stale. If it's still relevant to maintained versions, feel free to open a new issue
